File: magicSquare.java
public class magicSquare{
    
    static int[][] square;

    static int L, LIMIT, x=0, y=0, actual_val=1;

    public static void printSquare(){
        for(int i = 0; i < square.length; i++){
            for(int j = 0; j < square[i].length; j++){
                System.out.printf("%d\t",square[i][j]);
            }
            System.out.println();
        }
    }

    public static void first(){
        System.out.println("1. A matriz sempre terá os lados ímpares.");
    }

    public static void second(){
        System.out.println("2. O primeiro número é posicionada no meio da primeira coluna.");
        y = (int)(L/2);
        square[x][y] = actual_val;
        printSquare(); 
    }

    public static void third(){
        x =(x+((L*2)-1)) % L;
        y =(y+1) % L;

        if(square[x][y] != 0){
            fourth();
            return;     
        }

        System.out.println("3. O próximo número é colocado deslocando uma posição para cima \n"
                            +"e  uma posição para direita. Caso seja alcançada uma das \n"
                            +"bordas do quadrado, a posição é invertida.");
        square[x][y] = actual_val;
        printSquare();
    }

    public static void fourth(){
        y = (y+((L*2)-1)) % L;
        x = (x+2) % L;

        System.out.println("4. Caso já haja um número na nova posição, volta-se a posição antiga \n"
                            +"e apenas descemos linha. A coluna continua a mesma.");
        square[x][y] = actual_val;
        printSquare();
    }

    public static void main(String[] args){
        if(args.length == 1){
            L = Integer.parseInt(args[0]);

            if(L%2 != 0){
                square = new int[L][L];
                first();
                second();
                LIMIT = L*L;
                for(actual_val = 2; actual_val <= LIMIT; actual_val++){
                    third();
                }

            }else{
                System.out.println("FALHA: violação da regra 1, \n\t1. A matriz sempre terá os lados ímpares.:");
            }

        }else{
            System.out.println("Erro de argumentos: passe um numero impar para definir os lados do quadrado");
        }

    }
}

File: magicSquare.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

import sys
import numpy

if len(sys.argv) != 2 :
    sys.exit("Erro de argumentos: passe um numero impar para definir os lados do quadrado")
elif int(sys.argv[1])%2 != 1:
    sys.exit("FALHA: violação da regra 1, \n\t1. A matriz sempre terá os lados ímpares.")

x = 0
y = 0
L = int(sys.argv[1])
LIMIT = L*L
square = [[0 for i in range(L)] for i in range(L)]
actual_val = 1

################################################################
#DEFINITIONS
################################################################
def printSquare():
    s = ""
    for i in range(L):
        for j in range(L):
            s = s+'['+str(square[i][j])+']'
        s = s+'\n'
    print(s)

def first():
    print("1. A matriz sempre terá os lados ímpares.")
    printSquare()

def second():
    global y, x, actual_val, square
    y = L/2
    print("2. O primeiro número é posicionada no meio da primeira coluna.")
    square[x][y] = actual_val
    actual_val += 1
    printSquare()

def third():
    global x,y,square, actual_val
    x =(x+((L*2)-1))%L
    y =(y+1)%L
    if square[x][y] == 0 :
        print("3. O próximo número é colocado deslocando uma posição para cima \ne  uma posição para direita. Caso seja alcançada uma das \nbordas do quadrado, a posição é invertida.")       
        square[x][y] = actual_val
        actual_val += 1
        printSquare()
    else:
        fourth()

def fourth():
    global x,y,square, actual_val
    print("4. Caso já haja um número na nova posição, volta-se a posição antiga\n e apenas descemos linha. A coluna continua a mesma.");
    y = (y+((L*2)-1))%L;
    x = (x+2)%L;
    square[x][y] = actual_val
    actual_val += 1
    printSquare()

################################################################
#MAKE THE MAGIC
################################################################
first()

second()

while actual_val <= LIMIT :
    third()
